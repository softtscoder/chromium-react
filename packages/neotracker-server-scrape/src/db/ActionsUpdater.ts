import { addressToScriptHash, RawAction } from '@neo-one/client-full';
import { createChild, serverLogger } from '@neotracker/logger';
import { Action as ActionModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { strip0x } from '../utils';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface ActionsSaveSingle {
  readonly action: RawAction;
  readonly transactionID: string;
  readonly transactionHash: string;
}
export interface ActionsSave {
  readonly actions: ReadonlyArray<ActionsSaveSingle>;
}
export interface ActionsRevert {
  readonly transactionIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class ActionsUpdater extends SameContextDBUpdater<ActionsSave, ActionsRevert> {
  public async save(context: Context, { actions }: ActionsSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_actions' });
    await Promise.all(
      _.chunk(actions, context.chunkSize).map(async (chunk) =>
        ActionModel.insertAll(
          context.db,
          context.makeQueryContext(),
          chunk.map(({ action, transactionID, transactionHash }) => ({
            id: action.globalIndex.toString(),
            type: action.type,
            block_id: action.blockIndex,
            transaction_id: transactionID,
            transaction_hash: transactionHash,
            transaction_index: action.transactionIndex,
            index: action.index,
            script_hash: strip0x(addressToScriptHash(action.address)),
            // tslint:disable no-any no-null-keyword
            message: (action as any).message === undefined ? null : encodeURIComponent((action as any).message),
            args_raw: (action as any).args === undefined ? null : JSON.stringify((action as any).args),
            // tslint:disable-enable no-any no-null-keyword
          })),
        ),
      ),
    );
  }

  public async revert(context: Context, { transactionIDs }: ActionsRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_actions' });
    await Promise.all(
      _.chunk(transactionIDs, context.chunkSize).map(async (chunk) =>
        ActionModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('transaction_id', chunk)
          .delete(),
      ),
    );
  }
}
