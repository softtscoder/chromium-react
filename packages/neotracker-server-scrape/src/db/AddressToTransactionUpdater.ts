import { createChild, serverLogger } from '@neotracker/logger';
import { AddressToTransaction as AddressToTransactionModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressToTransactionSaveSingle {
  readonly addressIDs: ReadonlyArray<string>;
  readonly transactionID: string;
}
export interface AddressToTransactionSave {
  readonly transactions: ReadonlyArray<AddressToTransactionSaveSingle>;
}
export interface AddressToTransactionRevert {
  readonly transactionIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressToTransactionUpdater extends SameContextDBUpdater<
  AddressToTransactionSave,
  AddressToTransactionRevert
> {
  public async save(context: Context, { transactions }: AddressToTransactionSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_address_to_transaction' });
    const data = _.flatMap(transactions, ({ addressIDs, transactionID }) =>
      [...new Set(addressIDs)].map((addressID) => ({
        id1: addressID,
        id2: transactionID,
      })),
    );
    await Promise.all(
      _.chunk(data, context.chunkSize).map(async (chunk) => {
        await AddressToTransactionModel.insertAll(context.db, context.makeQueryContext(), chunk);
      }),
    );
  }

  public async revert(context: Context, { transactionIDs }: AddressToTransactionRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_address_to_transaction' });
    await Promise.all(
      _.chunk(transactionIDs, context.chunkSize).map((chunk) =>
        AddressToTransactionModel.query(context.db)
          .context(context.makeQueryContext())
          .delete()
          .whereIn('id2', chunk),
      ),
    );
  }
}
