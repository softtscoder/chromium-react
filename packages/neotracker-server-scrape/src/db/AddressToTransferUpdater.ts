import { createChild, serverLogger } from '@neotracker/logger';
import { AddressToTransfer as AddressToTransferModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressToTransferSaveSingle {
  readonly addressIDs: ReadonlyArray<string>;
  readonly transferID: string;
}
export interface AddressToTransferSave {
  readonly transfers: ReadonlyArray<AddressToTransferSaveSingle>;
}
export interface AddressToTransferRevert {
  readonly transferIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressToTransferUpdater extends SameContextDBUpdater<AddressToTransferSave, AddressToTransferRevert> {
  public async save(context: Context, { transfers }: AddressToTransferSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_address_to_transfer' });
    const data = _.flatMap(transfers, ({ addressIDs, transferID }) =>
      [...new Set(addressIDs)].map((addressID) => ({
        id1: addressID,
        id2: transferID,
      })),
    );
    await Promise.all(
      _.chunk(data, context.chunkSize).map(async (chunk) => {
        await AddressToTransferModel.insertAll(context.db, context.makeQueryContext(), chunk);
      }),
    );
  }

  public async revert(context: Context, { transferIDs }: AddressToTransferRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_address_to_transfer' });
    await Promise.all(
      _.chunk(transferIDs, context.chunkSize).map(async (chunk) => {
        await AddressToTransferModel.query(context.db)
          .context(context.makeQueryContext())
          .delete()
          .whereIn('id2', chunk);
      }),
    );
  }
}
