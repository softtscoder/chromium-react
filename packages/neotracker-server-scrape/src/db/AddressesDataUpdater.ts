import { createChild, serverLogger } from '@neotracker/logger';
import { Address as AddressModel, isPostgres } from '@neotracker/server-db';
import { raw } from 'objection';
import { Addresses, Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressesDataSave {
  readonly addresses: Addresses;
  readonly blockIndex: number;
  readonly blockTime: number;
}
export interface AddressesDataRevert {
  readonly addresses: Addresses;
  readonly transactionIDs: ReadonlyArray<string>;
  readonly blockIndex: number;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressesDataUpdater extends SameContextDBUpdater<AddressesDataSave, AddressesDataRevert> {
  public async save(context: Context, { addresses, blockIndex, blockTime }: AddressesDataSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_addresses_data' });
    await Promise.all(
      Object.entries(addresses).map(
        async ([address, { transactionCount, transferCount, transactionID, transactionHash }]) => {
          await AddressModel.query(context.db)
            .context(context.makeQueryContext())
            .where('id', address)
            .where('aggregate_block_id', '<', blockIndex)
            .patch({
              // tslint:disable no-any
              transaction_count: raw(`transaction_count + ${transactionCount}`) as any,
              transfer_count: raw(`transfer_count + ${transferCount}`) as any,
              // tslint:enable no-any
              last_transaction_id: transactionID,
              last_transaction_hash: transactionHash,
              last_transaction_time: blockTime,
              aggregate_block_id: blockIndex,
            });
        },
      ),
    );
  }

  public async revert(context: Context, { addresses, transactionIDs, blockIndex }: AddressesDataRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_addresses_data' });
    await Promise.all(
      Object.entries(addresses).map(async ([address, { transactionCount, transferCount }]) => {
        await AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .where('id', address)
          .where('aggregate_block_id', '>=', blockIndex)
          .patch({
            // tslint:disable no-any
            transaction_count: raw(`transaction_count - ${transactionCount}`) as any,
            transfer_count: raw(`transfer_count - ${transferCount}`) as any,
            // tslint:enable no-any
            aggregate_block_id: blockIndex - 1,
          });
      }),
    );

    const addressIDsSet = Object.keys(addresses);
    if (addressIDsSet.length > 0) {
      if (isPostgres(context.db)) {
        await context.db
          .raw(
            `
              WITH cte AS (
                SELECT a.id AS address_id, b.transaction_id, b.hash, b.block_time
                FROM address a
                LEFT OUTER JOIN (
                  SELECT address_id, transaction_id, hash, block_time
                  FROM (
                    SELECT id1 AS address_id, MAX(id2) AS transaction_id
                    FROM address_to_transaction
                    WHERE
                      id1 IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')}) AND
                      id2 NOT IN (${transactionIDs.map((id) => `'${id}'`).join(', ')})
                    GROUP BY address_to_transaction.id1
                  ) a
                  JOIN transaction b ON
                    a.transaction_id = b.id
                ) b ON
                  a.id = b.address_id
                WHERE
                  a.id IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')})
              )
              UPDATE address SET
                last_transaction_id=cte.transaction_id,
                last_transaction_hash=cte.hash,
                last_transaction_time=cte.block_time
              FROM cte
              WHERE
                id = cte.address_id
              `,
          )
          .queryContext(context.makeQueryContext());
      } else {
        await context.db
          .raw(
            `
              WITH cte AS (
                SELECT a.address_id, transaction_id, hash, block_time
                FROM (
                  SELECT id1 AS address_id, MAX(id2) AS transaction_id
                  FROM address_to_transaction
                  WHERE
                    id1 IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')}) AND
                    id2 NOT IN (${transactionIDs.map((id) => `'${id}'`).join(', ')})
                  GROUP BY id1
                ) a
                JOIN 'transaction' b ON
                  a.transaction_id = b.id
              )
              UPDATE address SET
                last_transaction_id=(select cte.transaction_id from cte where address_id = address.id),
                last_transaction_hash=(select cte.hash from cte where address_id = address.id),
                last_transaction_time=(select cte.block_time from cte where address_id = address.id)
              WHERE
                id IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')})
                `,
          )
          .queryContext(context.makeQueryContext());
      }
    }
  }
}
