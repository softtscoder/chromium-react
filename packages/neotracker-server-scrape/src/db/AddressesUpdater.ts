import { createChild, serverLogger } from '@neotracker/logger';
import { Address as AddressModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressesSave {
  readonly addresses: ReadonlyArray<Partial<AddressModel> & { readonly id: string }>;
}
export interface AddressesRevert {
  readonly addresses: ReadonlyArray<{ readonly id: string; readonly transactionID?: string }>;
  readonly blockIndex: number;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressesUpdater extends SameContextDBUpdater<AddressesSave, AddressesRevert> {
  public async save(context: Context, { addresses }: AddressesSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_addresses' });
    const existingAddressIDs = await Promise.all(
      _.chunk(addresses, context.chunkSize).map((chunk) =>
        AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('id', chunk.map(({ id }) => id)),
      ),
    ).then((result) => _.flatMap(result).map((address) => address.id));

    const existingAddressIDsSet = new Set(existingAddressIDs);
    const toCreate = addresses.filter((address) => !existingAddressIDsSet.has(address.id));

    await Promise.all(
      _.chunk(toCreate, context.chunkSize).map(async (chunk) =>
        AddressModel.insertAll(context.db, context.makeQueryContext(), chunk),
      ),
    );
  }

  public async revert(context: Context, { addresses, blockIndex }: AddressesRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_addresses' });
    const addressModels = await Promise.all(
      _.chunk(addresses, context.chunkSize).map((chunk) =>
        AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('id', chunk.map(({ id }) => id)),
      ),
    ).then((result) => _.flatMap(result));

    const addressToTransactionID = _.fromPairs(addresses.map(({ id, transactionID }) => [id, transactionID]));
    const toDelete = addressModels.filter(
      (address) =>
        (address.transaction_id == undefined && address.block_id === blockIndex) ||
        (address.transaction_id != undefined && address.transaction_id === addressToTransactionID[address.id]),
    );

    await Promise.all(
      _.chunk(toDelete, context.chunkSize).map((chunk) =>
        AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('id', chunk.map((address) => address.id))
          .delete(),
      ),
    );
  }
}
