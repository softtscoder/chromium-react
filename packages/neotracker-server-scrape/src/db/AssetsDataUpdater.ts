import { createChild, serverLogger } from '@neotracker/logger';
import { Asset as AssetModel } from '@neotracker/server-db';
import { raw } from 'objection';
import { Assets, Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AssetsDataSave {
  readonly assets: Assets;
  readonly blockIndex: number;
}
export interface AssetsDataRevert {
  readonly assets: Assets;
  readonly blockIndex: number;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AssetsDataUpdater extends SameContextDBUpdater<AssetsDataSave, AssetsDataRevert> {
  public async save(context: Context, { assets, blockIndex }: AssetsDataSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_assets' });
    await Promise.all(
      Object.entries(assets).map(async ([asset, { issued, transactionCount, addressCount, transferCount }]) => {
        await AssetModel.query(context.db)
          .context(context.makeQueryContext())
          .where('id', asset)
          .where('aggregate_block_id', '<', blockIndex)
          .patch({
            // tslint:disable no-any
            issued: raw(`issued + (${issued.toString()})`) as any,
            transaction_count: raw(`transaction_count + ${transactionCount}`) as any,
            address_count: raw(`address_count + ${addressCount}`) as any,
            transfer_count: raw(`transfer_count + ${transferCount}`) as any,
            // tslint:enable no-any
            aggregate_block_id: blockIndex,
          });
      }),
    );
  }

  public async revert(context: Context, { assets, blockIndex }: AssetsDataRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_assets' });
    await Promise.all(
      Object.entries(assets).map(async ([asset, { issued, transactionCount, addressCount, transferCount }]) => {
        await AssetModel.query(context.db)
          .context(context.makeQueryContext())
          .where('id', asset)
          .where('aggregate_block_id', '>=', blockIndex)
          .patch({
            // tslint:disable no-any
            issued: raw(`issued - (${issued.toString()})`) as any,
            transaction_count: raw(`transaction_count - ${transactionCount}`) as any,
            address_count: raw(`address_count - ${addressCount}`) as any,
            transfer_count: raw(`transfer_count - ${transferCount}`) as any,
            // tslint:enable no-any
            aggregate_block_id: blockIndex - 1,
          });
      }),
    );
  }
}
