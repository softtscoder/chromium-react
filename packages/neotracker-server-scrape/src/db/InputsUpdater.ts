import { createChild, serverLogger } from '@neotracker/logger';
import { TransactionInputOutput as TransactionInputOutputModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { InputUpdater } from './InputUpdater';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface InputsSaveSingle {
  readonly inputs: ReadonlyArray<TransactionInputOutputModel>;
  readonly transactionID: string;
  readonly transactionHash: string;
}
export interface InputsSave {
  readonly transactions: ReadonlyArray<InputsSaveSingle>;
  readonly blockIndex: number;
}
export interface InputsRevert {
  readonly references: ReadonlyArray<TransactionInputOutputModel>;
}
export interface InputsUpdaters {
  readonly input: InputUpdater;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class InputsUpdater extends SameContextDBUpdater<InputsSave, InputsRevert> {
  private readonly updaters: InputsUpdaters;

  public constructor(
    updaters: InputsUpdaters = {
      input: new InputUpdater(),
    },
  ) {
    super();
    this.updaters = updaters;
  }

  public async save(context: Context, { transactions, blockIndex }: InputsSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_inputs' });
    await Promise.all(
      _.flatMap(transactions, ({ inputs, transactionID, transactionHash }) =>
        inputs.map(async (reference) =>
          this.updaters.input.save(context, { reference, transactionID, transactionHash, blockIndex }),
        ),
      ),
    );
  }

  public async revert(context: Context, { references }: InputsRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_inputs' });
    await Promise.all(references.map(async (reference) => this.updaters.input.revert(context, { reference })));
  }
}
