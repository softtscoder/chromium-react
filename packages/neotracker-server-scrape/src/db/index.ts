export * from './BlockUpdater';
export * from './utils';
