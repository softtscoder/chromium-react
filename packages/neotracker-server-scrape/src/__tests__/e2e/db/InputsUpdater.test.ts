import { TransactionInputOutput } from '@neotracker/server-db';
import Knex from 'knex';
import { InputsRevert, InputsSave, InputsUpdater } from '../../../db/InputsUpdater';
import { data, makeContext } from '../../data';
import { updaterUnitTest, UpdaterUtilTestOptions } from '../../data/updaterTestUtil';

const inputOptions: ReadonlyArray<{
  readonly transactionID: string;
  readonly transactionHash: string;
  readonly blockIndex: number;
}> = [
  {
    transactionID: '12341234',
    transactionHash: '7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c',
    blockIndex: 10,
  },
  {
    transactionID: '23452345',
    transactionHash: '7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c',
    blockIndex: 10,
  },
];

const initiateTest = async (db: Knex) => {
  const context = makeContext({ db });

  const references = await Promise.all(
    inputOptions.map((input) =>
      TransactionInputOutput.query(db)
        .context(context.makeQueryContext())
        .insertAndFetch(
          data.createTransactionInputOutput({
            id: input.transactionID,
            output_block_id: input.blockIndex,
            output_transaction_hash: input.transactionHash,
            output_transaction_id: input.transactionID,
          }),
        ),
    ),
  );

  return { references };
};

const getInputs = (references: ReadonlyArray<TransactionInputOutput>): InputsSave => ({
  transactions: references.slice(0, 1).map((ref) => ({
    inputs: [ref],
    transactionID: ref.output_transaction_id,
    transactionHash: ref.output_transaction_hash,
  })),
  blockIndex: 10,
});

const getSecondaryInputs = (references: ReadonlyArray<TransactionInputOutput>): InputsSave => ({
  transactions: references.slice(1).map((ref) => ({
    inputs: [ref],
    transactionID: ref.output_transaction_id,
    transactionHash: ref.output_transaction_hash,
  })),
  blockIndex: 10,
});

const getReversions = (references: ReadonlyArray<TransactionInputOutput>): InputsRevert => ({
  references: references.slice(1),
});

const utilOptions: UpdaterUtilTestOptions<InputsSave, InputsRevert, ReadonlyArray<TransactionInputOutput>> = {
  name: `Inputs Updater`,
  initiateTest,
  createUpdater: () => new InputsUpdater(),
  getInputs,
  getSecondaryInputs,
  getReversions,
};

updaterUnitTest(utilOptions);
