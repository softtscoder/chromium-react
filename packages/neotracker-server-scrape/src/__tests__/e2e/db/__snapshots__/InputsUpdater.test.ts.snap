// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inputs Updater Inputs Updater: handles duplicate inserts/saves 1`] = `
Object {
  "action": Array [],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [
    Object {
      "address_id": "APyEx5f4Zm4oCHwFWiSTaph1fAAAAAAAAA",
      "asset_id": "g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049OOOOOOOOO",
      "claim_transaction_hash": null,
      "claim_transaction_id": null,
      "claim_value": "0",
      "id": "12341234",
      "input_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "input_transaction_id": "12341234",
      "output_block_id": 10,
      "output_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "output_transaction_id": "12341234",
      "output_transaction_index": 1,
      "subtype": "NONE",
      "type": "INPUT",
      "value": "1",
    },
    Object {
      "address_id": "APyEx5f4Zm4oCHwFWiSTaph1fAAAAAAAAA",
      "asset_id": "g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049OOOOOOOOO",
      "claim_transaction_hash": null,
      "claim_transaction_id": null,
      "claim_value": null,
      "id": "23452345",
      "input_transaction_hash": null,
      "input_transaction_id": null,
      "output_block_id": 10,
      "output_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "output_transaction_id": "23452345",
      "output_transaction_index": 1,
      "subtype": "NONE",
      "type": "INPUT",
      "value": "1",
    },
  ],
  "transfer": Array [],
}
`;

exports[`Inputs Updater Inputs Updater: handles insert/save 1`] = `
Object {
  "action": Array [],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [
    Object {
      "address_id": "APyEx5f4Zm4oCHwFWiSTaph1fAAAAAAAAA",
      "asset_id": "g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049OOOOOOOOO",
      "claim_transaction_hash": null,
      "claim_transaction_id": null,
      "claim_value": "0",
      "id": "12341234",
      "input_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "input_transaction_id": "12341234",
      "output_block_id": 10,
      "output_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "output_transaction_id": "12341234",
      "output_transaction_index": 1,
      "subtype": "NONE",
      "type": "INPUT",
      "value": "1",
    },
    Object {
      "address_id": "APyEx5f4Zm4oCHwFWiSTaph1fAAAAAAAAA",
      "asset_id": "g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049OOOOOOOOO",
      "claim_transaction_hash": null,
      "claim_transaction_id": null,
      "claim_value": null,
      "id": "23452345",
      "input_transaction_hash": null,
      "input_transaction_id": null,
      "output_block_id": 10,
      "output_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "output_transaction_id": "23452345",
      "output_transaction_index": 1,
      "subtype": "NONE",
      "type": "INPUT",
      "value": "1",
    },
  ],
  "transfer": Array [],
}
`;

exports[`Inputs Updater Inputs Updater: handles reverting inputs/saves 1`] = `
Object {
  "action": Array [],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [
    Object {
      "address_id": "APyEx5f4Zm4oCHwFWiSTaph1fAAAAAAAAA",
      "asset_id": "g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049OOOOOOOOO",
      "claim_transaction_hash": null,
      "claim_transaction_id": null,
      "claim_value": "0",
      "id": "12341234",
      "input_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "input_transaction_id": "12341234",
      "output_block_id": 10,
      "output_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "output_transaction_id": "12341234",
      "output_transaction_index": 1,
      "subtype": "NONE",
      "type": "INPUT",
      "value": "1",
    },
    Object {
      "address_id": "APyEx5f4Zm4oCHwFWiSTaph1fAAAAAAAAA",
      "asset_id": "g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049OOOOOOOOO",
      "claim_transaction_hash": null,
      "claim_transaction_id": null,
      "claim_value": null,
      "id": "23452345",
      "input_transaction_hash": null,
      "input_transaction_id": null,
      "output_block_id": 10,
      "output_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "output_transaction_id": "23452345",
      "output_transaction_index": 1,
      "subtype": "NONE",
      "type": "INPUT",
      "value": "1",
    },
  ],
  "transfer": Array [],
}
`;

exports[`Inputs Updater Inputs Updater: handles reverting inputs/saves 2`] = `
Object {
  "action": Array [],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [
    Object {
      "address_id": "APyEx5f4Zm4oCHwFWiSTaph1fAAAAAAAAA",
      "asset_id": "g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049OOOOOOOOO",
      "claim_transaction_hash": null,
      "claim_transaction_id": null,
      "claim_value": "0",
      "id": "12341234",
      "input_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "input_transaction_id": "12341234",
      "output_block_id": 10,
      "output_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "output_transaction_id": "12341234",
      "output_transaction_index": 1,
      "subtype": "NONE",
      "type": "INPUT",
      "value": "1",
    },
    Object {
      "address_id": "APyEx5f4Zm4oCHwFWiSTaph1fAAAAAAAAA",
      "asset_id": "g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049OOOOOOOOO",
      "claim_transaction_hash": null,
      "claim_transaction_id": null,
      "claim_value": "0",
      "id": "23452345",
      "input_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "input_transaction_id": "23452345",
      "output_block_id": 10,
      "output_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "output_transaction_id": "23452345",
      "output_transaction_index": 1,
      "subtype": "NONE",
      "type": "INPUT",
      "value": "1",
    },
  ],
  "transfer": Array [],
}
`;

exports[`Inputs Updater Inputs Updater: handles reverting inputs/saves 3`] = `
Object {
  "action": Array [],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [
    Object {
      "address_id": "APyEx5f4Zm4oCHwFWiSTaph1fAAAAAAAAA",
      "asset_id": "g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049OOOOOOOOO",
      "claim_transaction_hash": null,
      "claim_transaction_id": null,
      "claim_value": "0",
      "id": "12341234",
      "input_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "input_transaction_id": "12341234",
      "output_block_id": 10,
      "output_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "output_transaction_id": "12341234",
      "output_transaction_index": 1,
      "subtype": "NONE",
      "type": "INPUT",
      "value": "1",
    },
    Object {
      "address_id": "APyEx5f4Zm4oCHwFWiSTaph1fAAAAAAAAA",
      "asset_id": "g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049OOOOOOOOO",
      "claim_transaction_hash": null,
      "claim_transaction_id": null,
      "claim_value": null,
      "id": "23452345",
      "input_transaction_hash": null,
      "input_transaction_id": null,
      "output_block_id": 10,
      "output_transaction_hash": "7f48028c38117ac9e42c8e1f6f06ae027cdbb904eaf1a0bdc30c9d81694e045c",
      "output_transaction_id": "23452345",
      "output_transaction_index": 1,
      "subtype": "NONE",
      "type": "INPUT",
      "value": "1",
    },
  ],
  "transfer": Array [],
}
`;
